sql: use of constraint columns in FK can differ from order in PK definition

samxsmith opened the pull request:
sql: use of constraint columns in FK can differ from order in PK definition

ColumnIDs: Implement PermutationOf
- This methods returns true if the content of the two sets of ColumnIDs
is the same, even if they are not in the same order.
Contrary to the existing Equal() method, which also requires order
equality.

descpb: IndexDescriptor uses PermutationOf for comparing columns
- Previously when referring to an existing constraint, with a set of columns,
the Equals() method was used. That method requires order equality to match.
- Postgres recognises a constraint by reference even when the
provided fields are not in the order they were given when creating it.
- PermutationOf compares the column elements regardless of order.
This means that the behaviour of a constraint reference matches that of
Postgres.

Fixes #63324

Release note (sql change): The use order of columns in a foreign key no longer needs
to match the order the columns were defined for the reference table's unique constraint.

samxsmith committed:
sql: use of constraint columns in FK can differ from order in PK definition

ColumnIDs: Implement PermutationOf
- This methods returns true if the content of the two sets of ColumnIDs
is the same, even if they are not in the same order.
Contrary to the existing Equal() method, which also requires order
equality.

descpb: IndexDescriptor uses PermutationOf for comparing columns
- Previously when referring to an existing constraint, with a set of columns,
the Equals() method was used. That method requires order equality to match.
- Postgres recognises a constraint by reference even when the
provided fields are not in the order they were given when creating it.
- PermutationOf compares the column elements regardless of order.
This means that the behaviour of a constraint reference matches that of
Postgres.

Release note (sql change): The use order of columns in a foreign key no longer needs
to match the order the columns were defined for the reference table's unique constraint.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=65209) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=65209) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=65209) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65209)
<!-- Reviewable:end -->

otan commented on pkg/sql/catalog/descpb/structured.go:
```diff
@@ -145,6 +145,28 @@ func (c ColumnIDs) Equals(input ColumnIDs) bool {
+		return false
+	}
+
+	columnsLookup := map[ColumnID]bool{}
```
let's use `util.MakeFastIntSet` here. it is ok to cast the ColumnID to int.

samxsmith replied:
Have swapped to using this.
I think I have used it effectively but please let me know if there is a better way to make use of the Set.

otan commented on pkg/sql/catalog/descpb/structured.go:
```diff
@@ -145,6 +145,28 @@ func (c ColumnIDs) Equals(input ColumnIDs) bool {
 }
 
+// ElementWiseEquals returns true if the input list's elements are the same as this list's elements
+func (c ColumnIDs) ElementWiseEquals(input ColumnIDs) bool {
```
nitpick: `WiseEquals` is a bit vague for me.
we have "Equivalent" elsewhere, maybe that's the right name to use.

samxsmith replied:
Changed

otan commented on pkg/sql/catalog/descpb/structured_test.go:
```diff
@@ -0,0 +1,57 @@
+
+import "testing"
+
+func TestColumnIDs_ElementWiseEquals(t *testing.T) {
```
nit: `TestColumnIDsEquivalent`

samxsmith replied:
Changed

otan commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -1812,3 +1812,25 @@ CREATE TABLE duplicate_index_test (k INT PRIMARY KEY, v INT, INDEX idx (v));
 statement error pgcode 42P07 duplicate index name: \"idx\"
 ALTER TABLE duplicate_index_test ADD CONSTRAINT idx UNIQUE (v)
+
+##
```
nit: delete this line

samxsmith replied:
Deleted

otan commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -1812,3 +1812,25 @@ CREATE TABLE duplicate_index_test (k INT PRIMARY KEY, v INT, INDEX idx (v));
+
+# https://github.com/cockroachdb/cockroach/issues/63324
+# Referencing an existing constraint by fields in a different order
+# - should still find existing constraint
```
nit: remove `-`, end with full stop.

samxsmith replied:
Changed

otan commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -1812,3 +1812,25 @@ CREATE TABLE duplicate_index_test (k INT PRIMARY KEY, v INT, INDEX idx (v));
+# - should still find existing constraint
+
+statement ok
+create table reference_constraint_different_order_parent (a int not null, b int not null, c int not null, d varchar(10));
```
nit: remove trailing `;`. we can also one line the setup to be a bit more clear:

```
statement ok
create table reference_constraint_different_order_parent (a int not null, b int not null, c int not null, d varchar(10));
create table reference_constraint_different_order_child (id int, a int, b int, c int, e varchar(10));
alter table reference_constraint_different_order_parent add constraint p_pk primary key (a, b, c)

statement ok
alter table reference_constraint_different_order_child add constraint p_fk foreign key (a, b, c) references reference_constraint_different_order_parent (a, c, b)


```

Then can you add a test of the FK succeeding / failing?

samxsmith replied:
Have matched the format you've suggested

otan replied:
i would delete it from here

otan reviewed (commented):
Thanks! This looks close!

Few other things:
* Make sure each release note say `Release note: None`, or combine commits into one.
* I would combine the commits into one.

samxsmith commented:
@otan thank you for taking a look. I've made the changes you suggested.
Let me know if there's any more I can do.

otan commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -3671,3 +3671,17 @@ ALTER TABLE child_59582 ADD FOREIGN KEY (i) REFERENCES child_59582(rowid)
+alter table reference_constraint_different_order_parent add constraint p_pk primary key (a, b, c)
+
+statement ok
+alter table reference_constraint_different_order_child add constraint p_fk foreign key (a, b, c) references reference_constraint_different_order_parent (a, c, b)
```
ah, can you actually create table elements and check that referencing fails if the FK is/is not present in `reference_constraint_different_order_parent`?

postamar replied:
Could you reformat the SQL statements you added in the logic tests using https://sqlfum.pt/ please?

postamar reviewed (commented):
Thanks for contributing! Please reference the issue number in the message body of the appropriate commit. You can do this by adding the line "Fixes #63324."

No further comments beyond those outstanding by @otan.

samxsmith commented on pkg/sql/create_table.go:
```diff
@@ -1078,6 +1083,28 @@ func ResolveFK(
+// findTargetConstraintColumnIDs returns columnIDs as they are defined on the target table.
+// It is important to use them as defined on that table, so we use them in the order they are defined.
+// Otherwise, references created with columns defined out of order will fail foreign key checks.
+func findTargetConstraintColumnIDs(targetUniqueConstraint descpb.UniqueConstraint, target *tabledesc.Mutable) []ColumnID {
```
Further testing raised this point to me. In order to use a referenced index, it's necessary that the columnIDs on the FK definition are in the same order as on the target table. FK Checks fail otherwise. Rather than changing the FK checks in each place they were used, it seemed easier to change them here, once, at the point of definition.

I'm not very keen on this implementation, but I don't see a better way. Perhaps you have some ideas about other ways that we can get the data we need?
We could extend `descpb.UniqueConstraint` for instance, to include a method that returns the target constraint definition?

otan replied:
i'm not opposed to making the constraints return the definition if you think it makes it cleaner.

samxsmith replied:
`descpb.UniqueConstraint` now returns ColumnIDs. This did require adding the method to all found implementations, but they all had the ColumnIDs field so each is a very simple getter.

otan commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -3671,3 +3671,52 @@ ALTER TABLE child_59582 ADD FOREIGN KEY (i) REFERENCES child_59582(rowid)
+ALTER TABLE reference_constraint_different_order_child
+	ADD CONSTRAINT p_fk FOREIGN KEY (a, b, c) REFERENCES reference_constraint_different_order_parent (a, c, b);
+
+statement ok
```
can you insert something that is not in the same order and check that it is errors? look up `statement error` as per above.

samxsmith replied:
The INSERT shouldn't error. It's the foreign key field order that is the issue

otan replied:
yes, i mean add an entry that would violate the foreign key and prove that it errors with `statement error`

samxsmith replied:
Ahhh sorry I misunderstood, will add

otan reviewed (commented):
looks like some lint errors + CI failures as well to address.
you'll also need to rebase

otan commented:
(adding someone from queries to look at the FK changes too, since it may affect the optimiser output / why your tests are failing)

samxsmith commented:
> looks like some lint errors + CI failures as well to address.
> you'll also need to rebase

Trying to understand the test errors with `testccllogic` atm, but making little progress

mgartner reviewed (changes requested):
Thanks for the contribution! 

I'd squash all this into a single commit. Make sure to follow the commit message guidelines [here](https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages). Notably, the title of the commit should have the package name(s) where the changes took place. In this case `sql: ...` is good. Also, can you expand the release note a bit, maybe something like: `Release note (sql change): The order of reference columns in a foreign key no longer needs to match the order of columns in the reference table's unique constraint.`

I'll look into the `testccllogic` failures and report back.

> Reviewed 2 of 3 files at r12, 5 of 5 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-:-Maimbm65SfCx4XnmA0O:b-y99aml)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @postamar, and @samxsmith)

___
*[pkg/sql/catalog/table_elements.go, line 173 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-Maig6QA-N1SXNN05AEr:-Maig6QA-N1SXNN05AEs:b-5da9v5) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/catalog/table_elements.go#L173)):*
> ```Go
> 	GetCompositeColumnID(compositeColumnOrdinal int) descpb.ColumnID
> 
> 	GetColumnIDs() descpb.ColumnIDs
> ```

nit: move this up, below `GetColumnID(columnOrdinal int) descpb.ColumnID`
___
*[pkg/sql/catalog/descpb/structured.go, line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-M_giuqqkTEnBs38laIt-r3-149:-Maij5oo0ItkMnrdSuPz:b-rf19qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/5976917e6fcc6a443dd745a8268e6cc0fda8a73c/pkg/sql/catalog/descpb/structured.go#L149)):*
<details><summary><i>Previously, samxsmith (Sam X Smith) wrote…</i></summary><blockquote>

Changed
</blockquote></details>

A name that better captures the essence of the function might be `PermutationOf`?
___
*[pkg/sql/catalog/descpb/structured.go, line 149 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MaigWEc0iA6ok85JNI-:-MaigWEc0iA6ok85JNI0:b-e2si80) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/catalog/descpb/structured.go#L149)):*
> ```Go
> }
> 
> // Equivalent returns true if the input list's elements are the same as this list's elements
> ```

nit: comment lines should be <= 80 characters, and this should end with a period.
___
*[pkg/sql/catalog/descpb/structured.go, line 167 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MaijQElAnyilBwfRjD9:-MaijQElAnyilBwfRjDA:b-99fri1) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/catalog/descpb/structured.go#L167)):*
> ```Go
> 	}
> 
> 	return columnsLookup.Len() <= 0
> ```

Might be simpler to add both lists of columns into 2 sets in the same for loop (because lengths are the same after the initial check), and calling `FastIntSet.Equals`.

Note that your solution and this one I'm suggesting are only valid because duplicate columns are not allowed in indexes. If they were, this function would return false with two column lists like `(a, b, a, c)` and `(a, a, b, c)`. Can you add a note in the function comment explaining that this function will return false if either list of column ids has a duplicate?
___
*[pkg/sql/catalog/descpb/structured.go, line 407 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MailBKxAF5Jrk3fP8qz:-MailBKxAF5Jrk3fP8r-:b-jf6d36) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/catalog/descpb/structured.go#L407)):*
> ```Go
> }
> 
> // GetColumnIDs returns the constraint's column ids.
> ```

nit: this looks inadvertent
___
*[pkg/sql/catalog/descpb/structured.go, line 412 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MailJD37TXMIacvgk5R:-MailJD37TXMIacvgk5S:b-sg8woq) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/catalog/descpb/structured.go#L412)):*
> ```Go
> }
> 
> func (u *UniqueWithoutIndexConstraint) GetColumnIDs() ColumnIDs {
> ```

nit: add comment
___
*[pkg/sql/catalog/descpb/structured_test.go, line 46 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MailaHGCGvL8UIXsAZ5:-MailaHGCGvL8UIXsAZ6:btdtrjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/catalog/descpb/structured_test.go#L46)):*
> ```Go
> 			columnIDsTwo:   ColumnIDs{ColumnID(3), ColumnID(2), ColumnID(1)},
> 			expectedResult: true,
> 		},
> ```

nit: you could add tests with duplicate column IDs in either list and expect `False`, with a `name` to indicate that this is expected behavior for now.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 3690 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-Mailt3w4x6NFz1Ejb7-:-Mailt3w4x6NFz1Ejb70:bjbm2i1) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/logictest/testdata/logic_test/fk#L3690)):*
> ```
> 		INT8 NOT NULL,
> 	d
> 		VARCHAR(10)
> ```

nit: put the types on the same lines as the column names and use spaces instead of tabs
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 3693 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-Maim-Fu7wCWZuBbxHQg:-Maim-Fu7wCWZuBbxHQh:b762o6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/logictest/testdata/logic_test/fk#L3693)):*
> ```
> );
> CREATE TABLE reference_constraint_different_order_child (
> 	id INT8, a INT8, b INT8, c INT8, e VARCHAR(10)
> ```

nit: spaces instead of tabs
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 3696 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-Maim1qAEyLi282w5orB:-Maim1qAEyLi282w5orC:b-45twps) ([raw file](https://github.com/cockroachdb/cockroach/blob/12a30a211bf25db8e24d742ee7b7a84d501fafb3/pkg/sql/logictest/testdata/logic_test/fk#L3696)):*
> ```
> );
> ALTER TABLE reference_constraint_different_order_parent
> 	ADD CONSTRAINT p_pk PRIMARY KEY (a, b, c);
> ```

nit: spaces instead of tabs

You can also define the PK in the `CREATE TABLE` statement, e.g.:

```
CREATE TABLE t (
  a INT,
  b INT,
  c INT,
  PRIMARY KEY (a, b, c)
)
```


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
sorry forgot to submit this pending comment.
hopefully marcus can figure out the opt issue :)

mgartner commented:
The failing logictestccl tests were the result of adding implicit partition columns of a unique index to the `descpb.ForeignKeyConstraint.ReferencedColumnIDs` field. This made `ReferencedColumnIDs` contain more IDs than `OriginColumnIDs`, which is invalid and causes errors.

I think the best fix is to replace `GetColumnIDs` with `GetExplicitColumnIDs`. This commit shows the change that should work: https://github.com/mgartner/cockroach/commit/fe746b76a32da7933be8f6ed421da86cadcbc6da. @otan Does that look good to you?

@samxsmith Feel free to apply the diff directly to your branch. I also added a test in the `partitioning_implicit` test file which would have made this a bit more obvious to diagnose. You should include that test too.

mgartner commented:
Also, we should address this too: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/catalog/tabledesc/table.go#L369-L373. I can do that in a follow-up PR. Just leaving a note here as a reminder.

samxsmith commented:
@mgartner thank you for your review, and sorry for the delay in responding.

- I have squashed my changes into a single commit
- Moved the GetColumnIDs() so it is below GetColumnID(...)
- Renamed the Equivalent method `PermuationOf`
- Refactored that method to use two `MakeFastIntSets` as you suggest. With your suggestion, `PermutationOf` returns true with duplicates. I've added a test to demonstrate that.
- Replace the IsPartial comment I overwrote while clearly not looking properly, instead adding to the method below as intended.

samxsmith commented on pkg/sql/catalog/descpb/structured.go:
```diff
@@ -379,7 +404,11 @@ func (u *UniqueWithoutIndexConstraint) GetName() string {
 }
 
-// IsPartial returns true if the constraint is a partial unique constraint.
+// GetColumnIDs returns the constraint's column ids.
```
Yep sorry stupid mistake !

samxsmith commented:
@mgartner Thank you for your fix for the logictestccl tests. I've included those changes, including the test with implicit columns.
That test fails though with the following:
```
expected:
            CREATE TABLE public.fk_using_implicit_columns_against_t (
               pk INT8 NOT NULL,
               ref_t_pk INT8 NOT NULL,
               ref_t_c INT8 NOT NULL,
               CONSTRAINT "primary" PRIMARY KEY (pk ASC),
               CONSTRAINT fk_ref_t_pk_ref_t FOREIGN KEY (ref_t_pk) REFERENCES public.t(pk),
               CONSTRAINT fk_ref_t_c_ref_t FOREIGN KEY (ref_t_c) REFERENCES public.t(c),
               FAMILY fam_0_ref_t_pk_pk_ref_t_c (ref_t_pk, pk, ref_t_c)
            )

        but found (query options: "") :
            CREATE TABLE public.fk_using_implicit_columns_against_t (
               pk INT8 NOT NULL,
               ref_t_pk INT8 NOT NULL,
               ref_t_c INT8 NOT NULL,
               CONSTRAINT "primary" PRIMARY KEY (pk ASC),
               CONSTRAINT fk_ref_t_pk_ref_t FOREIGN KEY (ref_t_pk) REFERENCES public.t(pk),
               CONSTRAINT fk_ref_t_c_ref_t FOREIGN KEY (ref_t_c) REFERENCES public.t(c),
               FAMILY fam_0_ref_t_pk (ref_t_pk),
               FAMILY fam_1_pk (pk),
               FAMILY fam_2_ref_t_c (ref_t_c)
            )
```

mgartner reviewed (commented):
> Reviewed 6 of 6 files at r11, 8 of 8 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65209)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @postamar, and @samxsmith)

___
*[pkg/sql/catalog/table_elements.go, line 173 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-Maig6QA-N1SXNN05AEr:-MbN1xBS9r5I24aLfdUC:b-51hmb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/da07e638fe59f974f914e1edeeaf0964a1bc1948/pkg/sql/catalog/table_elements.go#L173)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: move this up, below `GetColumnID(columnOrdinal int) descpb.ColumnID`
</blockquote></details>

nit: I think you meant to move this up, but might have moved back down due to my patch (sorry about that).
___
*[pkg/sql/catalog/descpb/structured.go, line 149 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbN2Mmu3CjQSb47OXkJ:-MbN2Mmu3CjQSb47OXkK:bd0c4le) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cac4a5c23f64952c6f1173234a41ccee0f0a0e7/pkg/sql/catalog/descpb/structured.go#L149)):*
> ```Go
> }
> 
> // PermutationOf returns true if the input elements are the same as this list's.
> ```

We should add a note here to make it clear what the behavior is for duplicate IDs. Something like:

```
// PermutationOf returns true if this list and the input list contain the same
// set of column IDs in any order. If either list has a duplicate column ID,
// this function returns true if there is at least one instance of the column ID
// in the other list. For example:
// 
//     ColumnIDs{1, 2, 3, 1}.PermutationOf(ColumnIDs{1, 2, 3})
//       => true
//
//     ColumnIDs{1, 2, 3}.PermutationOf(ColumnIDs{1, 2, 3, 1}) => true
//       => true
//
```
___
*[pkg/sql/catalog/descpb/structured_test.go, line 52 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbN16Ua8_7slu5a9_6y:-MbN16Ub0y2s9AT7Atw-:b-uaifuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cac4a5c23f64952c6f1173234a41ccee0f0a0e7/pkg/sql/catalog/descpb/structured_test.go#L52)):*
> ```Go
> 			columnIDsTwo:   ColumnIDs{ColumnID(3), ColumnID(1), ColumnID(2), ColumnID(1)},
> 			expectedResult: true,
> 		},
> ```

I was thinking of some tests like the ones below, because the behavior might be surprising to people.

```
{
  name:           "duplicate in first list",
  columnIDsOne:   ColumnIDs{ColumnID(1), ColumnID(2), ColumnID(1), ColumnID(3)},
  columnIDsTwo:   ColumnIDs{ColumnID(3), ColumnID(2), ColumnID(1)},
  expectedResult: true,
}, 
{
  name:           "duplicate in second list",
  columnIDsOne:   ColumnIDs{ColumnID(2), ColumnID(1), ColumnID(3)},
  columnIDsTwo:   ColumnIDs{ColumnID(3), ColumnID(1), ColumnID(2), ColumnID(1)},
  expectedResult: true,
}
```


<!-- Sent from Reviewable.io -->

samxsmith commented:
@mgartner Ahh yes, I've move the method back up and added the comment you suggested.
Also added the extra duplicate tests for PermutationOf.

No progress on the CC failures yet though. Any ideas?

mgartner reviewed (changes requested):
> Reviewed 2 of 3 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65209)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @otan, @postamar, and @samxsmith)

___
*[pkg/sql/catalog/descpb/structured.go, line 155 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbNWJpJ26f1J99pIaV2:-MbNWJpJ26f1J99pIaV3:ba5yf5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/af590e864d3c5d92071f9d04ee79d7173eadfb0f/pkg/sql/catalog/descpb/structured.go#L155)):*
> ```Go
> //
> //     ColumnIDs{1, 2, 3, 1}.PermutationOf(ColumnIDs{1, 2, 3})
> //       => false
> ```

Both these examples return true, no?

I think I confused us because I incorrectly stated that the `FastIntSet.Equals` method would be the same as your original.

But I think your original algorithm behaved like this:

```
ColumnIDs{1, 2, 3, 1}.PermutationOf(ColumnIDs{1, 2, 3, 1})
  => false

ColumnIDs{1, 2, 3, 1}.PermutationOf(ColumnIDs{1, 2, 3})
  => true

ColumnIDs{1, 2, 3}.PermutationOf(ColumnIDs{1, 2, 3, 1})
  => false
```

And this new version now behaves like this:

```
ColumnIDs{1, 2, 3, 1}.PermutationOf(ColumnIDs{1, 2, 3, 1})
  => true

ColumnIDs{1, 2, 3, 1}.PermutationOf(ColumnIDs{1, 2, 3})
  => true

ColumnIDs{1, 2, 3}.PermutationOf(ColumnIDs{1, 2, 3, 1})
  => true
```

I think this is fine, we just need to make sure the comments and test cases match.
___
*[pkg/sql/catalog/descpb/structured.go, line 157 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbNYfVkAH8vIXXOGr41:-MbNYfVkAH8vIXXOGr42:b-qu6i4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/af590e864d3c5d92071f9d04ee79d7173eadfb0f/pkg/sql/catalog/descpb/structured.go#L157)):*
> ```Go
> //       => false
> //
> //     ColumnIDs{1, 2, 3}.PermutationOf(ColumnIDs{1, 2, 3, 1}) => true
> ```

nit: remove `=> true` at the end (this was my typo)


<!-- Sent from Reviewable.io -->

mgartner commented:
> @mgartner Thank you for your fix for the logictestccl tests. I've included those changes, including the test with implicit columns.
> That test fails though with the following

@samxsmith Sorry about that test case not working! I think it's failing because if a column family is not explicitly set for a table, we random assign a set of column families. So this test is probably failing sporadically.

Replacing the test with the test below should do the trick:

```
statement ok
CREATE TABLE fk_using_implicit_columns_against_t_one_family (
  pk INT NOT NULL PRIMARY KEY,
  ref_t_pk INT NOT NULL REFERENCES t,
  ref_t_c INT NOT NULL REFERENCES t(c),
  FAMILY (pk, ref_t_pk, ref_t_c)
)

query T
SELECT create_statement FROM [SHOW CREATE TABLE fk_using_implicit_columns_against_t_one_family]
----
CREATE TABLE public.fk_using_implicit_columns_against_t_one_family (
   pk INT8 NOT NULL,
   ref_t_pk INT8 NOT NULL,
   ref_t_c INT8 NOT NULL,
   CONSTRAINT "primary" PRIMARY KEY (pk ASC),
   CONSTRAINT fk_ref_t_pk_ref_t FOREIGN KEY (ref_t_pk) REFERENCES public.t(pk),
   CONSTRAINT fk_ref_t_c_ref_t FOREIGN KEY (ref_t_c) REFERENCES public.t(c),
   FAMILY fam_0_pk_ref_t_pk_ref_t_c (pk, ref_t_pk, ref_t_c)
)
```

mgartner commented:
@samxsmith also, you'll need to rebase off of `cockroachdb/cockroach@master` - looks like there some conflicts.

samxsmith commented:
@mgartner I've updated the comment to read as below. These patch the test inputs and outcomes.

```
// PermutationOf returns true if this list and the input list contain the same
// set of column IDs in any order. Returns false if either list has a duplicate
// which is not matched by the other list. For example:
//
//     ColumnIDs{2, 1, 3}.PermutationOf(ColumnIDs{3, 1, 2, 1})
//       => false
//
//     ColumnIDs{1, 2, 1, 3}.PermutationOf(ColumnIDs{3, 2, 1})
//       => false
//
//     However
//     ColumnIDs{1, 2, 1, 3}.PermutationOf(ColumnIDs{3, 1, 2, 1})
//     => true
```

mgartner reviewed (changes requested):
Thanks for the update. I appreciate you responding to all the comments. This is getting closer!

I've left a few more comments. You'll also need to rebase your branch on the latest `cockroachdb/cockroach@master` branch because there are merged confilcts.

> Reviewed 1 of 2 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-:-MbhT18f-MhlzDmTb2AM:brjq8rs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @otan, @postamar, and @samxsmith)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit, line 224 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbhNcpKDZ3EKeSuzwQK:-MbhNcpKDZ3EKeSuzwQL:b-7ie18d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddcc703622b73fa584abec7f51325113a2ec6c0/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L224)):*
> ```
> # Tests for FK references against the implicit columns.
> statement ok
> CREATE TABLE fk_using_implicit_columns_against_t_one_family (
> ```

You'll have to leave `CREATE TABLE fk_using_implicit_columns_against_t (...)` as-is because tests below rely on it. `CREATE TABLE fk_using_implicity_columns_against_t_one_family (...)` should be its own table.
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit, line 243 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbhNsvy-m0P_x4U0jqU:-MbhNsvy-m0P_x4U0jqV:b-6mq056) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddcc703622b73fa584abec7f51325113a2ec6c0/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L243)):*
> ```
>    FAMILY fam_0_pk_ref_t_pk_ref_t_c (pk, ref_t_pk, ref_t_c)
> )
> ```

nit: remove extra new line
___
*[pkg/sql/catalog/descpb/structured.go, line 166 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbhQVfGC1eV6RigZ-Q8:-MbhQVfGC1eV6RigZ-Q9:b-j81qg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddcc703622b73fa584abec7f51325113a2ec6c0/pkg/sql/catalog/descpb/structured.go#L166)):*
> ```Go
> 	if len(input) != len(c) {
> 		return false
> 	}
> ```

I was confused and missed this length check. Sorry about that.

`ColumnIDs{1, 1, 2}.PermutationOf(ColumnIDs{1, 2, 2})` will still return `true` even with the length check (can you add this as a test case too?). This is not consistent with the function description: `Returns false if either list has a duplicate which is not matched by the other list.`

Otherwise, we're only handling very special cases of duplicates. I'd rather we are explicit about this function's limitations than fix a few special cases and potentially surprise users of the functions, or we make the function behave as expected with duplicates.

So I think we should do one of two things:

1. Remove the length check and return `true` for all the duplicate column ID examples above. 
2. Or, Correctly handle all duplicate column ID cases by sorting each list and checking `c[i] == input[i]` for all `i`.

I'm now leaning towards **2** given all the confusion we've already had over this function.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 3704 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbhP7AHCqNxyLFcqE2L:-MbhP7AHCqNxyLFcqE2M:b5k5i0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddcc703622b73fa584abec7f51325113a2ec6c0/pkg/sql/logictest/testdata/logic_test/fk#L3704)):*
> ```
> INSERT
> INTO
> 	reference_constraint_different_order_parent
> ```

nit: you can remove the tabs and move the lines with them up, here and below. for example:

```diff
- VALUES
-         (1, 2, 3, 'par_val')
+ VALUES (1, 2, 3, 'par_val')
```
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 3711 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/65209#-MbhS-KW9gmgJzVUYjVt:-MbhS-KW9gmgJzVUYjVu:b-dal8ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddcc703622b73fa584abec7f51325113a2ec6c0/pkg/sql/logictest/testdata/logic_test/fk#L3711)):*
> ```
> 	reference_constraint_different_order_child
> VALUES
> 	(99, 1, 2, 3, 'child_val');
> ```

This test looks broken. The FK on child `(a, b, c)` references the parent `(a, c, b)`. You're inserting `(a, b, c) = (1, 2, 3)` into the child, but there is no parent row with `(a, c, b) = (1, 2, 3)`. So this test should be a FK violation error, not a success, correct?

As Oliver suggested, you should have a test that succeeds and one that fails. [Here](https://github.com/cockroachdb/cockroach/blob/9ddcc703622b73fa584abec7f51325113a2ec6c0/pkg/sql/logictest/testdata/logic_test/fk#L23) is an example to follow for testing a FK violation error.


<!-- Sent from Reviewable.io -->

samxsmith commented:
@mgartner I've just done the rebase, I think I got it right. `ColumnIDs` seems to have changed to `KeyColumnIDs` on the `IndexDescriptor`.

samxsmith commented:
@mgartner Ahh sorry I misunderstood the use of the CREATE TABLE. I've put it back, and added the new one separately.

I took a look at the PermutationsOf with fresh eyes and I too realised I'd been missing the length check there. I'll remove that, like you suggest, to avoid these inconsistent responses. The result is that duplicates now have no impact on outcome, as with the original implementation. I've also added a test of the same length with different duplicates as you raised.

mgartner reviewed (changes requested):
Nice work! We're getting closer.

mgartner commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -3671,3 +3671,51 @@ ALTER TABLE child_59582 ADD FOREIGN KEY (i) REFERENCES child_59582(rowid)
+INTO
+	reference_constraint_different_order_child
+VALUES
+	(99, 1, 2, 3, 'child_val');
```
This insert should not be successful, it should be a FK violation. There is no row in the parent where `(a, c, b) = (1, 2, 3)`. I think the issue is that we are setting `ReferencedColumnIDs` to the result of `targetDesc.GetExplicitColumnIDs()` without ordering these columns according to the column ordering of the `ALTER TABLE .. ADD CONSTRAINT .. FOREIGN KEY` statement.

Feel free to try to solve this, if you'd like. We should have tests for both a successful INSERT and a failed INSERT due to a FK violation. I'm also happy to take over the PR from here.

samxsmith replied:
I now realise that my assumption that this should work was why I went off on the tangent of using the existing PK definition with `GetExplicitColumnIDs()` in the first place. But you're right, it shouldn't work, because the reference is `b -> c` and vice versa, not `b -> b` etc.

Refreshing that assumption, I believe I was actually done a while back, before I went on this tangent.
I will fix the test expectations, then remove the unnecessary changes.

samxsmith replied:
If indeed that is the intention?

samxsmith replied:
@mgartner subject to that, I've added a test which checks the correct behaviour, and one which proves a violation otherwise. Let me know if those tests match what you were expecting 😄

samxsmith commented:
Am getting the following error in Bazel CI:
```
pkg/sql/colexec/colexec_test_/colexec_test
    exited with error code 142
```

otan reviewed (approved):
thanks @samxsmith ! it needed a rebase, i've done that for you. your name's still on the commit. 

bors r+

samxsmith commented:
Wow! Thanks everyone for your help & patience with my first contribution here.
@otan @mgartner @postamar 🙏

mgartner commented:
Thanks for the contribution!
